### PR TITLE
Add disable download source configuration

### PR DESCRIPTION
**Pull request by lunny**
Add configuration to enable/disable download source from UI.

### CODE REVIEW

File: custom/conf/app.example.ini [hunk 1]
```diff
@@ -878,6 +878,9 @@ ROUTER = console
 ;; Allow deletion of unadopted repositories
 ;ALLOW_DELETION_OF_UNADOPTED_REPOSITORIES = false
 
+;; Dont' allow download source tar files from UI
+;DISABLED_DOWNLOAD_SOURCE = false
```
Comment (techknowlogick) [inline custom/conf/app.example.ini]: Please also add to cheat sheet docs

Comment (techknowlogick) [reply custom/conf/app.example.ini]: I think you should mention the source *archive* not just source

**Review by 6543 — COMMENTED** [review]
https://github.com/go-gitea/gitea/blob/main/docs/content/doc/advanced/config-cheat-sheet.en-us.md patch missing

File: custom/conf/app.example.ini (hunk 1 above)
Comment (kittenswolf) [reply custom/conf/app.example.ini]: small typo - `Dont'`

**Comment by lunny** [comment]
@techknowlogick @kittenswolf @6543 Done.

**Review by 6543 — APPROVED** [review]

File: modules/setting/repository.go [hunk 2]
```diff
@@ -48,6 +48,7 @@ var (
 		DefaultBranch                           string
 		AllowAdoptionOfUnadoptedRepositories    bool
 		AllowDeleteOfUnadoptedRepositories      bool
+		DisabledDownloadSourceArchives          bool
```
Comment (zeripath) [inline modules/setting/repository.go]:
```suggestion
		DisableDownloadSourceArchives          bool
```

Reads better and is more consistent with the rest of the options here

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 3]
```diff
@@ -78,6 +78,7 @@ Values containing `#` or `;` must be quoted using `` ` `` or `"""`.
 - `DEFAULT_BRANCH`: **main**: Default branch name of all repositories.
 - `ALLOW_ADOPTION_OF_UNADOPTED_REPOSITORIES`: **false**: Allow non-admin users to adopt unadopted repositories
 - `ALLOW_DELETION_OF_UNADOPTED_REPOSITORIES`: **false**: Allow non-admin users to delete unadopted repositories
+- `DISABLED_DOWNLOAD_SOURCE_ARCHIVES`: **false**: Don't allow download source archive files from UI
```
Comment (zeripath) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]:
```suggestion
- `DISABLE_DOWNLOAD_SOURCE_ARCHIVES`: **false**: Don't allow download source archive files from UI
```

File: custom/conf/app.example.ini [hunk 4]
```diff
@@ -878,6 +878,9 @@ ROUTER = console
 ;; Allow deletion of unadopted repositories
 ;ALLOW_DELETION_OF_UNADOPTED_REPOSITORIES = false
 
+;; Don't allow download source archive files from UI
+;DISABLED_DOWNLOAD_SOURCE_ARCHIVES = false
```
Comment (zeripath) [inline custom/conf/app.example.ini]:
```suggestion
;DISABLE_DOWNLOAD_SOURCE_ARCHIVES = false
```

File: modules/templates/base.go [hunk 5]
```diff
@@ -35,10 +35,11 @@ func BaseVars() Vars {
...
 
-		"ShowRegistrationButton":      setting.Service.ShowRegistrationButton,
-		"ShowMilestonesDashboardPage": setting.Service.ShowMilestonesDashboardPage,
-		"ShowFooterBranding":          setting.ShowFooterBranding,
-		"ShowFooterVersion":           setting.ShowFooterVersion,
+		"ShowRegistrationButton":         setting.Service.ShowRegistrationButton,
+		"ShowMilestonesDashboardPage":    setting.Service.ShowMilestonesDashboardPage,
+		"ShowFooterBranding":             setting.ShowFooterBranding,
+		"ShowFooterVersion":              setting.ShowFooterVersion,
+		"DisabledDownloadSourceArchives": setting.Repository.DisabledDownloadSourceArchives,
```
Comment (zeripath) [inline modules/templates/base.go]:
```suggestion
		"DisableDownloadSourceArchives": setting.Repository.DisableDownloadSourceArchives,
```

File: routers/web/web.go [hunk 6]
```diff
@@ -290,6 +290,13 @@ func RegisterRoutes(m *web.Route) {
 		}
 	}
 
+	dlSourceEnabled := func(ctx *context.Context) {
+		if setting.Repository.DisabledDownloadSourceArchives {
```
Comment (zeripath) [inline routers/web/web.go]:
```suggestion
		if setting.Repository.DisableDownloadSourceArchives {
```

File: templates/mail/release.tmpl [hunk 7]
```diff
@@ -31,12 +31,14 @@
 		<br>
 		{{.locale.Tr "mail.release.downloads"}}
 		<ul>
+			{{if not .DisabledDownloadSourceArchives}}
```
Comment (zeripath) [inline templates/mail/release.tmpl]:
```suggestion
			{{if not .DisableDownloadSourceArchives}}
```

File: templates/repo/branch/list.tmpl [hunk 8]
```diff
@@ -26,13 +26,15 @@
 										{{svg "octicon-git-branch"}}
 									</div>
 								{{end}}
-								<div class="ui basic jump dropdown icon button tooltip" data-content="{{$.locale.Tr "repo.branch.download" ($.DefaultBranch)}}" data-position="top right">
-									{{svg "octicon-download"}}
-									<div class="menu">
-										<a class="item archive-link" href="{{$.RepoLink}}/archive/{{PathEscapeSegments $.DefaultBranch}}.zip" rel="nofollow">{{svg "octicon-file-zip"}}&nbsp;ZIP</a>
-										<a class="item archive-link" href="{{$.RepoLink}}/archive/{{PathEscapeSegments $.DefaultBranch}}.tar.gz" rel="nofollow">{{svg "octicon-file-zip"}}&nbsp;TAR.GZ</a>
+								{{if not $.DisabledDownloadSourceArchives}}
```
Comment (zeripath) [inline templates/repo/branch/list.tmpl]:
```suggestion
								{{if not $.DisableDownloadSourceArchives}}
```

File: templates/repo/branch/list.tmpl [hunk 9]
```diff
@@ -26,13 +26,15 @@
...
+									<div class="ui basic jump dropdown icon button tooltip" data-content="{{$.locale.Tr "repo.branch.download" ($.DefaultBranch)}}" data-position="top right">
+										{{svg "octicon-download"}}
+										<div class="menu">
+											<a class="item archive-link" href="{{$.RepoLink}}/archive/{{PathEscapeSegments $.DefaultBranch}}.zip" rel="nofollow">{{svg "octicon-file-zip"}}&nbsp;ZIP</a>
+											<a class="item archive-link" href="{{$.RepoLink}}/archive/{{PathEscapeSegments $.DefaultBranch}}.tar.gz" rel="nofollow">{{svg "octicon-file-zip"}}&nbsp;TAR.GZ</a>
+										</div>
 									</div>
-								</div>
+								{{end}}
```
Comment (zeripath) [inline templates/repo/branch/list.tmpl:37]: Is there a missing close div now?

File: templates/repo/home.tmpl [hunk 10]
```diff
@@ -124,6 +124,7 @@
 				{{if eq $n 0}}
 					<div class="ui action tiny input" id="clone-panel">
 						{{template "repo/clone_buttons" .}}
+						{{if not .DisabledDownloadSourceArchives}}
```
Comment (zeripath) [inline templates/repo/home.tmpl]:
```suggestion
						{{if not .DisableDownloadSourceArchives}}
```

File: templates/repo/release/list.tmpl [hunk 11]
```diff
@@ -37,8 +37,10 @@
 								<div class="download df ac">
 									{{if $.Permission.CanRead $.UnitTypeCode}}
 										<a class="mr-3 mono" href="{{$.RepoLink}}/src/commit/{{.Sha1}}" rel="nofollow">{{svg "octicon-git-commit" 16 "mr-2"}}{{ShortSha .Sha1}}</a>
-										<a class="archive-link mr-3" href="{{$.RepoLink}}/archive/{{.TagName | PathEscapeSegments}}.zip" rel="nofollow">{{svg "octicon-file-zip" 16 "mr-2"}}ZIP</a>
-										<a class="archive-link mr-3" href="{{$.RepoLink}}/archive/{{.TagName | PathEscapeSegments}}.tar.gz" rel="nofollow">{{svg "octicon-file-zip" 16 "mr-2"}}TAR.GZ</a>
+										{{if not $.DisabledDownloadSourceArchives}}
```
Comment (zeripath) [inline templates/repo/release/list.tmpl]:
```suggestion
										{{if not $.DisableDownloadSourceArchives}}
```

File: templates/repo/release/list.tmpl [hunk 12]
```diff
@@ -104,8 +106,10 @@
 							<div class="download">
 							{{if $.Permission.CanRead $.UnitTypeCode}}
 								<a class="mono" href="{{$.RepoLink}}/src/commit/{{.Sha1}}" rel="nofollow">{{svg "octicon-git-commit" 16 "mr-2"}}{{ShortSha .Sha1}}</a>
-								<a class="archive-link" href="{{$.RepoLink}}/archive/{{.TagName | PathEscapeSegments}}.zip" rel="nofollow">{{svg "octicon-file-zip"}}&nbsp;ZIP</a>
-								<a class="archive-link" href="{{$.RepoLink}}/archive/{{.TagName | PathEscapeSegments}}.tar.gz" rel="nofollow">{{svg "octicon-file-zip"}}&nbsp;TAR.GZ</a>
+								{{if not $.DisabledDownloadSourceArchives}}
```
Comment (zeripath) [inline templates/repo/release/list.tmpl]:
```suggestion
								{{if not $.DisableDownloadSourceArchives}}
```

File: templates/repo/release/list.tmpl [hunk 13]
```diff
@@ -146,7 +150,7 @@
 									{{$.locale.Tr "repo.release.downloads"}}
 								</summary>
 								<ul class="list">
-									{{if and (not .IsDraft) ($.Permission.CanRead $.UnitTypeCode)}}
+									{{if and (not $.DisabledDownloadSourceArchives) (not .IsDraft) ($.Permission.CanRead $.UnitTypeCode)}}
```
Comment (zeripath) [inline templates/repo/release/list.tmpl]:
```suggestion
									{{if and (not $.DisableDownloadSourceArchives) (not .IsDraft) ($.Permission.CanRead $.UnitTypeCode)}}
```

File: templates/repo/branch/list.tmpl (hunk 9 above)
Comment (lunny) [reply templates/repo/branch/list.tmpl:37]: I don't think so. There are still two div before end. GH's git diff highlight make you confusing.

**Comment by lunny** [comment]
@zeripath done.

**Review by zeripath — APPROVED** [review]

**Comment by 6543** [comment]
.

**Pull request merged**